src/rom_workflows.cc

```diff
@@ -429,11 +465,99 @@ void testROMPoissonOperator(MGmolInterface *mgmol_)
+    Mesh* mymesh             = Mesh::instance();
+    const pb::PEenv& myPEenv = mymesh->peenv();
+
+    if (ct.isLocMode())
+        printf("LocMode is On!\n");
+    else
+        printf("LocMode is Off!\n");
+
+    ROMPrivateOptions rom_options = ct.getROMOptions();
+
+    /* Load MGmol pointer and Potentials */
+    MGmol<OrbitalsType> *mgmol = static_cast<MGmol<OrbitalsType> *>(mgmol_);
+    Poisson *poisson = mgmol->electrostat_->getPoissonSolver(); 
+    Potentials& pot = mgmol->getHamiltonian()->potential();
+    std::shared_ptr<Rho<OrbitalsType>> rho = mgmol->getRho();
+    const OrthoType ortho_type = rho->getOrthoType();
+    assert(ortho_type == OrthoType::Nonorthogonal);
+
+    const int dim = pot.size();
+    printf("pot size: %d\n", dim);
+
+    const int nrows = mymesh->locNumpt();
+    printf("mesh::locNumpt: %d\n", nrows);
+
+    OrbitalsType *orbitals = mgmol->getOrbitals();
+    ProjectedMatricesInterface *proj_matrices = orbitals->getProjMatrices();
+    SquareLocalMatrices<MATDTYPE, memory_space_type>& localX(proj_matrices->getLocalX());
+
+    printf("localX nmat: %d\n", localX.nmat());
+    printf("localX n: %d\n", localX.n());
+    printf("localX m: %d\n", localX.m());
+
+    bool dm_distributed = (localX.nmat() > 1);
+    assert(!dm_distributed);
+    /* copy density matrix */
+    CAROM::Matrix dm(localX.getRawPtr(), localX.m(), localX.n(), dm_distributed, true);
+}
+
+/*
+    dm: density matrix converted to CAROM::Matrix
+    phi_basis: POD basis matrix of orbitals, or orbitals themselves
+    rom_phi: ROM coefficients of POD Basis. If phi_basis is orbitals themselves, then simply an identity.
+    local_idx: sampled local grid indices on this processor.
+    sampled_rho: FOM density on sampled grid points.
+*/
+void computeRhoOnSamplePts(const CAROM::Matrix &dm,
+    const CAROM::Matrix &phi_basis, const CAROM::Matrix &rom_phi,
+    const std::vector<int> &local_idx, CAROM::Vector &sampled_rho)
+{
+    assert(!dm.distributed());
+    assert(sampled_rho.distributed() == phi_basis.distributed());
+
+    // this will be resized.
+    CAROM::Matrix sampled_phi(1, 1, phi_basis.distributed());
+
+    CAROM::mult(phi_basis, local_idx, rom_phi, sampled_phi);
+
+    /* same product as in computeRhoSubdomainUsingBlas3 */
+    CAROM::Matrix *product = sampled_phi.mult(dm);
+
+    sampled_rho.setSize(sampled_phi.numRows());
+    double *d_product = product->getData();
+    double *d_phi = sampled_phi.getData();
+    for (int d = 0; d < sampled_rho.dim(); d++)
+    {
+        double val = 0.0;
+        /* CAROM Matrices are row-major */
+        for (int c = 0; c < sampled_phi.numColumns(); c++, d_product++, d_phi++)
+            val += (*d_product) * (*d_phi);
+
+        sampled_rho(d) = val;
+    }
+
+    /*
+        TODO(kevin): need to figure out what these functions do.
+        and probably need to make ROM-equivalent functions with another hyper-reduction?
+    */
+    // gatherSpin();
```

Comment:
"spin" can be ignored for now. when spin is on, it is like having two replicated copy of everything, on two different sets of processors, and in this case rho is the sum of these two parts -> gather op